source/fuzz/transformation_add_dead_break.cpp

```diff
@@ -0,0 +1,289 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_add_dead_break.h"
+#include "source/opt/basic_block.h"
+#include "source/opt/ir_context.h"
+#include "source/opt/struct_cfg_analysis.h"
+
+namespace spvtools {
+namespace fuzz {
+
+using opt::BasicBlock;
+using opt::IRContext;
+using opt::Instruction;
+
+BasicBlock* TransformationAddDeadBreak::MaybeFindBlock(
+    IRContext* context, uint32_t maybe_block_id) {
+  auto inst = context->get_def_use_mgr()->GetDef(maybe_block_id);
+  if (inst == nullptr) {
+    // No instruction defining this id was found.
+    return nullptr;
+  }
+  if (inst->opcode() != SpvOpLabel) {
+    // The instruction defining the id is not a label, so it cannot be a block
+    // id.
+    return nullptr;
+  }
+  return context->cfg()->block(maybe_block_id);
+}
+
+bool TransformationAddDeadBreak::PhiIdsOk(IRContext* context,
+                                          BasicBlock* bb_from,
+                                          BasicBlock* bb_to) {
+  if (bb_from->IsSuccessor(bb_to)) {
+    // There is already an edge from |from_block_| to |to_block_|, so there is
+    // no need to extend OpPhi instructions.  Do not allow phi ids to be
+    // present. This might turn out to be too strict; perhaps it would be OK
+    // just to ignore the ids in this case.
+    return phi_ids_.empty();
+  }
+  // The break would add a previously non-existent edge from |from_block_| to
+  // |to_block_|, so we go through the given phi ids and check that they exactly
+  // match the OpPhi instructions in |to_block_|.
+  uint32_t phi_index = 0;
+  // An explicit loop, rather than applying a lambda to each OpPhi in |bb_to|,
+  // makes sense here because we need to increment |phi_index| for each OpPhi
+  // instruction.
+  for (auto& inst : *bb_to) {
+    if (inst.opcode() != SpvOpPhi) {
+      // The OpPhi instructions all occur at the start of the block; if we find
+      // a non-OpPhi then we have seen them all.
+      break;
+    }
+    if (phi_index == phi_ids_.size()) {
+      // Not enough phi ids have been provided to account for the OpPhi
+      // instructions.
+      return false;
+    }
+    // Look for an instruction defining the next phi id.
+    Instruction* phi_extension =
+        context->get_def_use_mgr()->GetDef(phi_ids_[phi_index]);
+    if (!phi_extension) {
+      // The id given to extend this OpPhi does not exist.
+      return false;
+    }
+    if (phi_extension->type_id() != inst.type_id()) {
+      // The instruction given to extend this OpPhi either does not have a type
+      // or its type does not match that of the OpPhi.
+      return false;
+    }
+    // Check whether the given id corresponds to a constant.  If so, that's
+    // fine.
+    bool found = false;
```

Comment:
Instead of searching for |phi_extension| in all module scope values, I'd suggest getting the basic block that defines it. If it has no block, then it should be module scope and safe. Otherwise, you can verify that that basic block dominates |bb_from|.

Reply:
Done.